src/schedule/zephyr_ll.c

```diff
@@ -0,0 +1,296 @@
+		}
+	}
+
+	/*
+	 * If the task has not been added, it means that it has the lowest
+	 * priority and should be added at the end of the list
+	 */
+	if (list == &sch->tasks)
+		list_item_append(&task->list, &sch->tasks);
+
+	spin_unlock(&sch->lock);
+
+	ret = domain_register(sch->ll_domain, task, &zephyr_ll_run, sch);
+	if (ret < 0)
+		tr_err(&ll_tr, "zephyr_ll_task_schedule: cannot register domain %d",
+		       ret);
+
+	return 0;
+}
+
+/* This is synchronous - after this returns the object can be destroyed! */
+static int zephyr_ll_task_free(void *data, struct task *task)
+{
+	struct zephyr_ll *sch = data;
+
+	switch (task->state) {
+	case SOF_TASK_STATE_INIT:
+	case SOF_TASK_STATE_CANCEL:
+	case SOF_TASK_STATE_FREE:
+		break;
+	default:
+		spin_lock(&sch->lock);
+		/* Now the task is guaranteed to not be running */
+		list_item_del(&task->list);
+		spin_unlock(&sch->lock);
+	}
+
+	rfree(task->priv_data);
+
+	return 0;
+}
+
+/* This seems to be asynchronous? */
+static int zephyr_ll_task_cancel(void *data, struct task *task)
+{
+	/*
+	 * FIXME: this can be called from an interrupt, also on a different core
+	 * Once we have moved running tasks to a different list, only the status
+	 * check there will be protected by the spin-lock. Then we can take the
+	 * spin-lock here too. Doing this now can potentially deadlock.
+	 */
+	task->state = SOF_TASK_STATE_CANCEL;
+
+	return 0;
+}
+
+/*
+ * FIXME: clarify rescheduling semantics. Currently only DMA trace is using it.
+ * The goal is to reschedule a periodic task with a relatively long period to
+ * run at an earlier time. But the current ll-scheduler based implementation
+ * doesn't even attempt to check whether the task is in fact scheduled to run at
+ * an earlier time.
+ * While our implementation runs all tasks under a spin-lock we could do various
+ * checks, but when we switch over to drop the spin-lock when running tasks,
+ * we'll have to come up with a robust implementation.
+ */
+static int zephyr_ll_task_reschedule(void *data, struct task *task, uint64_t start)
+{
+	struct zephyr_ll *sch = data;
+	struct list_item *list;
+	struct task *task_iter;
+	uint64_t time = domain_time(sch) +
+		sch->ll_domain->ticks_per_ms * start / 1000;
+
+	spin_lock(&sch->lock);
+
+	/* check to see if we are already scheduled */
+	list_for_item(list, &sch->tasks) {
+		task_iter = container_of(list, struct task, list);
+		/* found it, but it can also be currently running */
+		if (task_iter == task) {
+			if (task->start > time)
+				/* set start time */
+				task->start = time;
+			break;
+		}
+	}
+
+	spin_unlock(&sch->lock);
+
+	return 0;
+}
+
+/* Runs on secondary cores in their shutdown sequence */
+static void zephyr_ll_scheduler_free(void *data)
+{
+	struct zephyr_ll *sch = data;
+	struct task *task;
+	struct list_item *list, *tmp;
+
+	spin_lock(&sch->lock);
+
+	/*
+	 * FIXME: similar to zephyr_ll_task_reschedule() - what do we do when
+	 * the scheduling loop isn't holding the lock any more?
+	 */
```

Comment:
Is this LL tick running at this point or not ? (as this impacts the complexity)

Reply:
it should be, yes